src/main/scala/io/iohk/ethereum/vm/OpCode.scala

```diff
@@ -546,6 +592,127 @@ case object LOG3 extends LogOp(0xa3)
+      val gCap = gasCap(state, gas, gExtra)
+      if (endowment == 0) gCap else gCap + G_callstipend
+    }
+
+    lazy val result = {
+      val toAddr = Address(to)
+
+      val (account, world1) = this match {
+        case CALL =>
+          val w = state.world.transfer(state.ownAddress, toAddr, endowment)
+          val acc = w.getGuaranteedAccount(toAddr)
+          (acc, w)
+
+        case _ =>
+          val acc = state.world.getAccount(toAddr).getOrElse(Account.Empty)
+          (acc, state.world)
+      }
+
+      val (owner, caller) = this match {
+        case CALL => (toAddr, state.ownAddress)
+        case CALLCODE => (state.ownAddress, state.ownAddress)
+        case DELEGATECALL => (state.ownAddress, state.env.callerAddr)
+      }
+
+      val env = state.env.copy(
+        ownerAddr = owner,
+        callerAddr = caller,
+        inputData = inputData,
+        value = callValue,
+        program = Program(world1.getCode(account.codeHash)),
+        callDepth = state.env.callDepth + 1)
+
+      val context = state.context.copy(
+        env = env,
+        startGas = startGas,
+        world = world1)
+
+      VM.run(context)
+    }
+
+    val validCall =
+      state.env.callDepth < OpCode.MaxCallDepth &&
+      endowment <= state.ownAccount.balance
+
+    if (!validCall || result.error.isDefined) {
+      val stack2 = stack1.push(DataWord.Zero)
+      val gasAdjustment = if (validCall) BigInt(0) else -startGas
+
+      state
+        .withStack(stack2)
+        .withMemory(mem1)
+        .spendGas(gasAdjustment)
+        .step()
+
+    } else {
+      val stack2 = stack1.push(DataWord(1))
+      val output = result.returnData.take(outSize.intValue)
+      val mem2 = mem1.store(outAddr, result.returnData)
+
+      state
+        .spendGas(-result.gasRemaining)
+        .withStack(stack2)
+        .withMemory(mem2)
+        .withWorld(result.world)
+        .withAddressesToDelete(result.addressesToDelete)
+        .step()
+    }
+  }
+
+  protected def varGas(state: ProgramState): BigInt = {
+    val (Seq(gas, to, callValue, inAddr, inSize, outAddr, outSize), _) = getParams(state)
+    val endowment = if (this == DELEGATECALL) BigInt(0) else callValue.toBigInt
+
+    val memCostIn = calcMemCost(state.memory.size, inAddr, inSize)
+    val memCostOut = calcMemCost(state.memory.size, outAddr, outSize)
+    val memCost = memCostIn max memCostOut
+
+    // FIXME: these are calculated twice (for gas and exec), especially account existence. Can we do better?
+    val gExtra = gasExtra(state, endowment, Address(to))
+    val gCap = gasCap(state, gas, gExtra)
+    memCost + gCap + gExtra
+  }
+
+  private def getParams(state: ProgramState): (Seq[DataWord], Stack) = {
+    val (Seq(gas, to), stack1) = state.stack.pop(2)
+    val (value, stack2) = if (this == DELEGATECALL) DataWord(state.env.value) -> stack1 else stack1.pop
+    val (Seq(inAddr, inSize, outAddr, outSize), stack3) = stack2.pop(4)
```

Comment:
I already explained it here: https://github.com/input-output-hk/etc-client/pull/60#commitcomment-21038366